File: day_5/main.js
import * as fs from 'node:fs';
import {exit} from 'node:process';

const fname = './input.txt';
// const fname = './test.txt';

const input_r = fs.readFileSync(fname, 'utf-8');
const input = input_r.split('\r\n');
const rules = {};
const updates = [];

// FIRST CREATE INPUT, make dictionary of rules (reversed)
let i = 0;
while (input[i] != '') {
  let vals = input[i].split('|');
  vals = vals.map((val) => Number(val));
  if (!rules[vals[1]]) {
    rules[vals[1]] = [];
  }
  rules[vals[1]].push(vals[0]);
  i++;
}

i++;
while (input[i]) {
  let vals = input[i].split(',');
  updates.push(vals.map((val) => Number(val)));
  i++;
}

// Now check the orderings
const valid_orderings = [];
let orderings_to_fix = {};
updates.forEach((update) => {
  i = 0;
  let seen = {};
  let valid_ordering = true;

  seen = calculate_seen(update);

  for (let j = 0; j < update.length; j++) {
    let num = update[j];
    const num_rules = rules[num];
    num_rules?.forEach((rule) => {
      // Make sure index of rule is before our number, if not, bad
      if (seen[rule] > seen[num]) {
        valid_ordering = false;
        // Lets keep track of orderings to fix for part 2
        orderings_to_fix[update] = true;
      }
    });
  }

  if (valid_ordering) {
    console.log('===VALID===');
    valid_orderings.push(update);
    return;
  }
  console.log('===INVALID===');
});

// Now lets turn the orderings we need to fix into a [][] of numbers
orderings_to_fix = Object.keys(orderings_to_fix);
orderings_to_fix = orderings_to_fix.map((v) =>
  v.split(',').map((v2) => Number(v2))
);

// Loop over all orders we need to fix
for (let x = 0; x < orderings_to_fix.length; x++) {
  let seen = calculate_seen(orderings_to_fix[x]);

  // Look at every number
  for (let j = 0; j < orderings_to_fix[x].length; j++) {
    let num = orderings_to_fix[x][j];
    const num_rules = rules[num];
    let swapped = false;
    // Look at all rules for this number, are there any invalid orderings?
    num_rules?.forEach((rule) => {
      // If an invalid ordering, swap them
      if (!swapped && seen[rule] > seen[num]) {
        orderings_to_fix[x][seen[rule]] = num;
        orderings_to_fix[x][seen[num]] = rule;
        seen = calculate_seen(orderings_to_fix[x]);
        j = -1; // -1 since j gets incremented to 0 on loop end
        swapped = true;
        // Lets keep track of orderings to fix for part 2
      }
    });
  }
}

let puzzle_input = 0;
valid_orderings.forEach((update) => {
  const index = Math.floor(update.length / 2);
  puzzle_input += update[index];
});

let fixed_input = 0;
orderings_to_fix.forEach((update) => {
  const index = Math.floor(update.length / 2);
  fixed_input += update[index];
});

console.log('Solution: ', puzzle_input);
console.log('Fixed Solution: ', fixed_input);

function calculate_seen(order) {
  const seen = {};
  let i = 0;
  order.forEach((num) => {
    seen[num] = i;
    i++;
  });
  return seen;
}

File: day_5/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var rules map[int][]int = make(map[int][]int)
var updates [][]int = make([][]int,0)

var valid_orderings [][]int = make([][]int,0)
var orderings_to_fix [][]int = make([][]int, 0)

func main(){
	file, _ := os.Open("./input.txt")
	defer file.Close()
	scanner := bufio.NewScanner(file)

	// First scan all the rules
	for scanner.Scan(){
		if(scanner.Text() == ""){
			break
		}
		rule := strings.Split(scanner.Text(), "|")
		int0, _ := strconv.Atoi(rule[0])
		int1, _ := strconv.Atoi(rule[1])

		rules[int1] = append(rules[int1], int0)
	}

	// Then scan all the updates
	i := 0
	for scanner.Scan(){
		update := strings.Split(scanner.Text(),",")
		updates = append(updates, make([]int,0))
		for _, v_s := range update{
			v, _ := strconv.Atoi(v_s)
			updates[i] = append(updates[i], v)
		}
		i++
	}

	// Lets now check all of the orderings
	
	for _, update := range updates{
		is_valid := true;
		seen := calculate_seen(update)

		update_loop:
		for _, val := range update{
			number_rules := rules[val]
			for _, rule := range number_rules{
				// Is the index of a dependency greater than our current value?
				if seen[rule] > seen[val]{
					is_valid = false
					break update_loop
				}
			}
		}
		if is_valid{
			valid_orderings = append(valid_orderings, update)
			continue
		}
		orderings_to_fix = append(orderings_to_fix, update)
	}

	// Now lets fix the invalid orderings

	for x, update := range orderings_to_fix{
		seen := calculate_seen(update)
		for j := 0; j < len(update); j++{
			num := update[j]
			number_rules := rules[num]
			for _, rule := range number_rules{
				if seen[rule] > seen[num]{
					// Swap the values
					orderings_to_fix[x][seen[rule]] = num
					orderings_to_fix[x][seen[num]] = rule

					// Recalculate Seen
					seen = calculate_seen(update)
					j = -1;
					break
				}
			}
		}
	}

	valid_score := 0
	fixed_score := 0

	for _, order := range valid_orderings{
		len := len(order)
		len = len / 2
		valid_score += order[len]
	}
	for _, order := range orderings_to_fix{
		len := len(order)
		len = len / 2
		fixed_score += order[len]
	}

	fmt.Println("=======")
	fmt.Println(valid_score)
	fmt.Println(fixed_score)
}


func calculate_seen(update []int)map[int]int{
	seen := make(map[int]int)
	for i, value := range update{
		seen[value] = i
	}
	return seen
}